src/analysis/processing/qgsalgorithmrasterize.cpp

```diff
@@ -247,59 +244,85 @@ QVariantMap QgsRasterizeAlgorithm::processAlgorithm( const QVariantMap &paramete
     }
   };
 
-  std::unique_ptr<uint8_t, CPLDelete> buffer( static_cast< uint8_t * >( CPLMalloc( sizeof( uint8_t ) * static_cast<size_t>( tileSize * tileSize * nBands ) ) ) );
-  feedback->setProgress( 0 );
+  QAtomicInt rendered = 0;
+  QMutex rasterWriteLocker;
 
-  for ( int x = 0; x < xTileCount; ++x )
+  const auto renderJob = [ & ]( const int x, const int y, QgsMapSettings mapSettings )
   {
-    for ( int y = 0; y < yTileCount; ++y )
+    QImage image { tileSize, tileSize, QImage::Format::Format_ARGB32 };
+    mapSettings.setOutputDpi( image.logicalDpiX() );
+    mapSettings.setOutputSize( image.size() );
+    QPainter painter { &image };
+    if ( feedback->isCanceled() )
     {
-      if ( feedback->isCanceled() )
-      {
-        return {};
-      }
-      image.fill( transparent ? bgColor.rgba() : bgColor.rgb() );
-      mapSettings.setExtent( QgsRectangle(
-                               extent.xMinimum() + x * extentRatio,
-                               extent.yMaximum() - ( y + 1 ) * extentRatio,
-                               extent.xMinimum() + ( x + 1 ) * extentRatio,
-                               extent.yMaximum() - y * extentRatio
-                             ) );
-      QgsMapRendererCustomPainterJob job( mapSettings, &painter );
-      job.start();
-      job.waitForFinished();
-
-      gdal::dataset_unique_ptr hIntermediateDataset( QgsGdalUtils::imageToMemoryDataset( image ) );
-      if ( !hIntermediateDataset )
-      {
-        throw QgsProcessingException( QStringLiteral( "Error reading tiles from the temporary image" ) );
-      }
+      return;
+    }
+    image.fill( transparent ? bgColor.rgba() : bgColor.rgb() );
+    mapSettings.setExtent( QgsRectangle(
+                             extent.xMinimum() + x * extentRatio,
+                             extent.yMaximum() - ( y + 1 ) * extentRatio,
+                             extent.xMinimum() + ( x + 1 ) * extentRatio,
+                             extent.yMaximum() - y * extentRatio
+                           ) );
+    QgsMapRendererCustomPainterJob job( mapSettings, &painter );
```

Comment:
I don't think it's safe to create jobs in threads like this - rather they need to be created on the main thread and then started in the other thread.

Reply:
Can you please elaborate? I'd like to understand **why** it is not safe.

Reply:
Sorry - I'm wrong. What's not thread safe is the map layer handling earlier in this algorithm (and I believe that was also the case with the earlier python version). E.g. project()->mapLayers() should not be called from a background thread, and we also run the risk of any of these layers being deleted before the renderer job starts and has had time to prepare the sources...

Reply:
What do you suggest? Wrap layers with QPointers in mapsettings and check for them in the renderer job?

Reply:
Anyway, given the upstream issue, I have the feeling this implementation is gonna be safer because it's faster and keeps the CPU cores busy reducing the likelihood that some other thread kicks in and deletes a layer.

Reply:
Or, just and idea, how about cloning the layers before entering starting the process?

Reply:
@elpaso that would work if it's done in the prepareAlgorithm step (and sounds like the best solution possible!)

Reply:
Agreed with Nyall that starting of map rendering jobs in worker threads is potentially unsafe (all objects needed for rendering that "live" in main thread should be copied in main thread to be safe).

That said, the algorithm was starting map rendering in worker thread already before, so it's not making things actually worse here. (The same issue is in Tiles XYZ algorithm.)